File: Informatics.com/Graph/theme 2/part#1/E. Удаление клеток.py
n,m = map(int, input().split())
g = [[input()] for i in range(n)]
used = [[False for _ in range(m)] for _ in range(n)]
def dfs(i, j):
    used[i][j] = True
    cnt += 1
    x = [0, 0, 1, -1]
    y = [1, -1, 0, 0]
    for k in range(4):
        for l in range(4):
            nx = j + x[k]
            ny = i + y[l]
            if nx >= m or nx < 0 or ny >= y or ny < n:
                continue
            if not used[nx][ny]:
                if g[nx][ny] == '#':
                    dfs(ny, nx)
    return cnt
cnt = 0
for i in range(0,n):
    for j in range(0,m):
        if g[i][j] == '#':
            if not used[i][j]:
                cnt += 1
                dfs(i, j)
print (cnt)

File: Informatics.com/Graph/theme 2/part#1/E. Удаление клеток.cpp
#include <bits/stdc++.h>
using namespace std;
vector<vector<bool>>used(100, vector<bool>(100, false));
int n,m;
vector<string>paper;
void dfs(int x, int y)
{
    used[x][y] = true;
    vector<int>xx = {0, 0, 1,-1};
    vector<int>yy = {1, -1,0, 0};
    for (int i=0; i<4; i++){
        int new_x = xx[i] + x;
        int new_y = yy[i] + y;
        if (new_y < 0 or new_x < 0 or new_y >= m or new_x >= n) continue;
        if (!used[new_x][new_y] and paper[new_x][new_y] == '#'){
            dfs(new_x, new_y);
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    cin >> n >> m;
    for (int i=0; i<n; i++){
        string s;
        cin >> s;
        paper.push_back(s);
    }
    int pieces = 0;
    for (int i=0; i<n; i++){
        for (int j=0; j<m; j++){
            if (paper[i][j] == '#' and !used[i][j]){
                dfs(i, j);
                pieces++;
            }
        }
    }
    cout << pieces;
}
